linkerd/client-policy/src/http/router.rs

```diff
@@ -0,0 +1,166 @@
+use super::Route;
+use crate::{Receiver, RoutePolicy};
+use futures::{future::MapErr, FutureExt, TryFutureExt};
+use linkerd_error::Error;
+use linkerd_stack::{layer, NewService, Oneshot, Param, Service, ServiceExt};
+use std::{
+    collections::{
+        hash_map::{self, HashMap},
+        HashSet,
+    },
+    sync::Arc,
+    task::{Context, Poll},
+};
+use tokio_util::sync::ReusableBoxFuture;
+
+/// A router that uses a `Service` that's built for each route policy (with a
+/// fallback service when no route is matched).
+///
+/// This router behaves similarly to the
+/// `linkerd_service_profiles::http::ServiceRouter`, but it operates on
+/// HTTPRoute policies rather than ServiceProfile policies.
+#[derive(Clone, Debug)]
+pub struct NewServiceRouter<N>(N);
+
+#[derive(Debug)]
+pub struct ServiceRouter<T, N, S> {
+    new_route: N,
+    target: T,
+    changed: ReusableBoxFuture<'static, Result<Receiver, Error>>,
+    http_routes: Arc<[Route]>,
+    services: HashMap<RoutePolicy, S>,
+    default: S,
+}
+
+// === impl NewServiceRouter ===
+
+impl<N> NewServiceRouter<N> {
+    pub fn layer() -> impl layer::Layer<N, Service = Self> + Clone {
+        layer::mk(Self)
+    }
+}
+
+impl<T, N> NewService<T> for NewServiceRouter<N>
+where
+    T: Param<Option<Receiver>> + Clone,
+    N: NewService<(Option<RoutePolicy>, T)> + Clone,
+{
```

Comment:
Do we really need to handle these as optional? We should use the stack to enforce that these are present; and *all* requests must have routes that go through this service, so I'd rather not promulgate optionality to the inner stack.

Reply:
I know this is different from how it's handled in service profiles; but I think we want to drive towards a state where there is a route description on all requests.

Reply:
the route is no longer optional as of fc32de7a3ea7c57229e5f716730203ae7b93ef71; the policy rx is still an `Option`, but i _could_ change this to operate on a separate target type (there's already a TODO about this)...